### PR TITLE
Feature: Open ai and note to flashcard

### PR DESCRIPTION
**Pull request by Roshan-y**
# Description

This PR introduces the `OpenAI` and `NotesToFlashcard` classes, which integrate OpenAI API capabilities for converting note content into flashcards. The `OpenAI` class sends requests to the OpenAI API, while `NotesToFlashcard` manages the process of generating flashcards from notes by utilizing OpenAI’s text completion functionality. The PR also includes comprehensive unit tests to validate the functionality of these classes, particularly the process of parsing API responses and creating `Flashcard` objects. 

Also refactored Scanner and OCR to utils file.

# How Has This Been Tested?

### Unit Tests
- **OpenAI Request Handling**: Ensures that `OpenAI` successfully constructs and sends requests to the API and properly handles responses and errors.
- **NotesToFlashcard JSON Parsing**: Validates that `NotesToFlashcard` correctly parses JSON responses, extracts flashcard data, and instantiates `Flashcard` objects.
- **Mock Integration with FlashcardRepository**: Verifies that `NotesToFlashcard` interacts with `FlashcardRepository` as expected, using Mockito to capture and assert arguments.

These tests can be reproduced by running the `NotesToFlashcardTest` class. 

# Dependencies

- **OkHttp**: Used for making HTTP requests to the OpenAI API.
- **Gson**: Used for constructing and parsing JSON data within the API request and response.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged in downstream modules, no conflicts with main

### CODE REVIEW

File: app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+   *
+   * @param note the note to be converted into flashcards
+   */
+  fun convertNoteToFlashcards(
+      note: Note,
+      onSuccess: (List<Flashcard>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    try {
+      // TODO: When markdown support is implemented, fetch the markdown file content using
```
Comment (robertaleG) [inline app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt:34]: nice comment

Comment (robertaleG) [reply app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt:34]: very informative

File: app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+          val answer = flashcardObject.get("answer").asString
+
+          val flashcard =
+              Flashcard(
+                  id = flashcardViewModel.getNewUid(),
+                  front = question,
+                  back = answer,
+                  nextReview = Timestamp.now(),
+                  userId = note.userId,
+                  folderId = "",
```
Comment (robertaleG) [inline app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt:96]: you can add the folderId of the notes, it is a field in Note.
replace it by `note.folderId?:""` as folderId is a nullable String.

File: app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+  }
+
+  /**
+   * Parses a JSON string response from the OpenAI API to extract flashcard data and converts it
+   * into a list of Flashcard objects.
+   *
+   * @param jsonResponse the JSON response from OpenAI API containing flashcard data
+   */
+  private fun parseFlashcardsFromJson(
```
Comment (robertaleG) [inline app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt:68]: very nice function

File: app/src/main/java/com/github/onlynotesswent/utils/OpenAI.kt [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+   * - finish_reason: Reason why the completion stopped (e.g., "stop", "length"). "stop" indicates
+   *   it completed the response naturally.
+   * - index: The position of the completion within the choices array.
+   *
+   * @param prompt the prompt to generate text from
+   * @param onSuccess the callback function to invoke on a successful response
+   * @param onFailure the callback function to invoke on a failed response
+   * @param model the model to use for generating text (default: "gpt-3.5-turbo")
+   */
+  fun sendRequest(
```
Comment (robertaleG) [inline app/src/main/java/com/github/onlynotesswent/utils/OpenAI.kt:50]: Bonus points for extra detailed documentation. 💯 well done

File: app/src/test/java/com/github/onlynotesswent/utils/NotesToFlashcardTest.kt [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  @Before
+  fun setup() {
+    MockitoAnnotations.openMocks(this)
+
+    // Initialize FirebaseApp with Robolectric context
+    FirebaseApp.initializeApp(org.robolectric.RuntimeEnvironment.getApplication())
+
+    // Mock FlashcardRepository and set up FlashcardViewModel with it
+    mockFlashcardRepository = mock(FlashcardRepository::class.java)
+    mockFlashcardViewModel = FlashcardViewModel(mockFlashcardRepository)
```
Comment (robertaleG) [inline app/src/test/java/com/github/onlynotesswent/utils/NotesToFlashcardTest.kt:96]: technically you don't need to call the view model mock, and the repository is already mocked with the openMocks on line 89. not that important

**Review by robertaleG — CHANGES_REQUESTED** [review]
Super nice PR: clean, well written, fully tested (I like that!) 

Merge main into it and you're good to go. I added some comments you should take a look at.
Particularly, I want you to add folderId to the flashcard instead of the empty string "", now that folders are implemented on main. For that small but important change, I will request changes but notify me as soon you've made the change and I'll quickly approve from my phone.

The rest is fantastic 🥇

File: app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt (hunk 2 above)
Comment (Roshan-y) [reply app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt:96]: Yes nice catch

File: app/src/main/java/com/github/onlynotesswent/utils/OpenAI.kt [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+          addProperty("content", prompt)
+        }
+
+    val messagesArray = JsonArray().apply { add(messageObject) }
+
+    val jsonBody =
+        JsonObject().apply {
+          addProperty("model", model)
+          add("messages", messagesArray)
+          addProperty("temperature", 0.2)
```
[the remark this answers is not in the document]
Comment (Roshan-y) [reply app/src/main/java/com/github/onlynotesswent/utils/OpenAI.kt:69]: The temperature adjusts the creativity of openai, so how much stuff he can make up I will add this as a comment on the file.

File: app/src/test/java/com/github/onlynotesswent/utils/NotesToFlashcardTest.kt [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+      assertEquals("Cryptocurrency examples", flashcards[2].front)
+      assertEquals(
+          "There are thousands of cryptocurrencies. Some of the best known include: Bitcoin, Ethereum...",
+          flashcards[2].back)
+
+      // Verify for each flashcard
+      for (flashcard in flashcards) {
+        assertEquals(testNote.userId, flashcard.userId)
+        assertEquals(testNote.id, flashcard.noteId)
```
[the remark this answers is not in the document]
Comment (Roshan-y) [reply app/src/test/java/com/github/onlynotesswent/utils/NotesToFlashcardTest.kt:131]: yes thank you

**Review by robertaleG — APPROVED** [review]
Nice jobs making changes in response to my review. I approve

**Pull request merged**